### PR TITLE
NN & MDX

### PR DESCRIPTION
**Pull request by romanpresniakov**

### CODE REVIEW

File: src/schnetpack/nn/cutoff.py
Comment (epens94) [inline src/schnetpack/nn/cutoff.py:1]: This part is not necessary. In schnetpack firstly the cutoff function is defined and called within the appropiate cutoff function class forward call. 
If one wants to use a specific cutoff function this is invoked in the config file as (e.g):
        cutoff_fn:
          _target_: schnetpack.nn.cutoff.CosineCutoff

File: src/schnetpack/nn/embed.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    Add offsets to distance vectors given a cell and cell offsets. Cell is assumed to be
+    Args:
+        r_ij (Array): Distance vectors, shape: (n_pairs,3)
+        cell (Array): Unit cell matrix, shape: (3,3). Unit cell vectors are assumed to be row-wise.
+        cell_offsets (Array): Offsets for each pairwise distance, shape: (n_pairs,3).
+    Returns:
+    """
+    offsets = np.einsum('...i, ij -> ...j', cell_offsets, cell)
+    return r_ij + offsets
+
```
Comment (epens94) [inline src/schnetpack/nn/embed.py]: The add cell offset is handled differently in schnetpack and efficiently implemented in this [class](https://github.com/atomistic-machine-learning/schnetpack/blob/337fa3f803725201b5890ec4ee07da0d4774dec8/src/schnetpack/transform/neighborlist.py#L219)

File: src/schnetpack/nn/embed.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+def _init_sphc_zeros(z, sph_ij, *args, **kwargs):
```
Comment (epens94) [inline src/schnetpack/nn/embed.py]: we can keep it but please make sure to use torch instead of numpy,

File: src/schnetpack/nn/embed.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    return {'chi': np.zeros((z.shape[-1], sph_ij.shape[-1]), dtype=sph_ij.dtype)}
+
+
+def _init_sphc(z, sph_ij, phi_r_cut, idx_i, point_mask, mp_normalization, *args, **kwargs):
```
Comment (epens94) [inline src/schnetpack/nn/embed.py]: we can keep it but please rewrite so that all comments in this PR are considered (safe_scale, torch instead of numpy, point_mask)

File: src/schnetpack/nn/embed.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+            else:
+                # Initialize SPHCs with a neighborhood dependent embedding
+                geometric_data.update(_init_sphc(z=z,
+                                                 sph_ij=sph_harms_ij,
+                                                 phi_r_cut=phi_r_cut,
+                                                 idx_i=idx_i,
+                                                 point_mask=point_mask,
+                                                 mp_normalization=self._lambda)
+                                      )
+
```
Comment (epens94) [inline src/schnetpack/nn/embed.py]: I would recommend to exclude all SPHC parts in a seperate module. Please make sure to use torch instead of numpy

File: src/schnetpack/nn/embed.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            rbf_ij = safe_scale(rbf_ij, scale=phi_r_cut[:, None])  # shape: (n_pairs,K)
+            g_ij = sph_harms_ij[:, :, None] * rbf_ij[:, None, :]  # shape: (n_pair,m_tot,K)
+            g_ij = safe_scale(g_ij, scale=pair_mask[:, None, None], placeholder=0)  # shape: (n_pair,m_tot,K)
+            geometric_data.update({'g_ij': g_ij})
+
+        return geometric_data
+
+
+class AtomTypeEmbed(nn.Module):
+    def __init__(self, features: int, prop_keys: Dict, num_embeddings: int = 100,
```
Comment (epens94) [inline src/schnetpack/nn/embed.py]: Atomtype embedding is implemented in [here](https://github.com/atomistic-machine-learning/schnetpack/blob/337fa3f803725201b5890ec4ee07da0d4774dec8/src/schnetpack/nn/embedding.py#L152) or directly in the init of the [representation](https://github.com/atomistic-machine-learning/schnetpack/blob/337fa3f803725201b5890ec4ee07da0d4774dec8/src/schnetpack/representation/schnet.py#L133-L134)

File: src/schnetpack/nn/embed.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        return safe_scale(nn.Embedding(num_embeddings=self.num_embeddings, embedding_dim=self.features)(z),
+                          scale=point_mask[:, None])
+
+    def __dict_repr__(self):
+        return {self.module_name: {'num_embeddings': self.num_embeddings,
+                                   'features': self.features,
+                                   'prop_keys': self.prop_keys}}
+
+
+class OneHotEmbed(nn.Module):
```
Comment (epens94) [inline src/schnetpack/nn/embed.py:12]: We can keep this.
Have you tested if this implementation and the original implementation result in the same initial one hot embedding ?
Please make sure to always use torch instead of numpy

File: src/schnetpack/nn/radial.py
Comment (epens94) [inline src/schnetpack/nn/radial.py:1]: This part is not necessary. 
If one wants to use a specific radial function this is invoked in the config file as (e.g):
radial_basis:
          _target_: schnetpack.nn.radial.GaussianRBF

File: src/schnetpack/nn/spherical.py
Comment (epens94) [inline src/schnetpack/nn/spherical.py:1]: We can discard this, since Its already implemented in a previously PR #1

File: src/schnetpack/properties.py
Comment (epens94) [inline src/schnetpack/properties.py:1]: point_mask and pair_mask is not necessary, since padding is not applied/needed in schnetpack
Those mask functionality is to ensure to only include indices of atoms that do not come from index padding.

File: src/schnetpack/utils/mask.py
Comment (epens94) [inline src/schnetpack/utils/mask.py:1]: I would recommend to postpone this functionality. We can check later if this is needed.
Please make sure that you use torch instead of numpy in function calls.

**Comment by epens94** [comment]
Hi Roman,

- The commit [36eb9ad](https://github.com/romanpresniakov/schnetpack/pull/4/commits/36eb9adda3460f30f16eaa82f953c7990c1ac469) is not necessary since all functionalities are covered [here](https://github.com/romanpresniakov/schnetpack/blob/master/src/schnetpack/md/simulation_hooks/callback_hooks.py)

- For the commit [76d5897](https://github.com/romanpresniakov/schnetpack/pull/4/commits/76d5897eaacc674b8304b8eb623a5b1660194c1c) Please see my inline comments. I would recommend to exclude all SPHC parts in a seperate module
